src/server/protocol.ts

```diff
@@ -546,7 +546,15 @@ namespace ts.server.protocol {
         command: CommandTypes.GetApplicableRefactors;
         arguments: GetApplicableRefactorsRequestArgs;
     }
-    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs;
+    export type GetApplicableRefactorsRequestArgs = FileLocationOrRangeRequestArgs & {
+        triggerReason?: RefactorTriggerReason;

```

Comment:
The simplest version of this is a boolean, `isUserInvoked`.  I can see why we might want to be able to generalize that in the future though, so moving to an enum, `{ Invoked, Speculative }`, seems reasonable - it gives us the option of addressing more nuances in the future.  I'm not sure why we'd go more abstract than that though - it seems unlikely we'll need both an enum and an additional payload.

Reply:
Switching to an enum 

```ts
export enum RefactorTriggerReason {
    Implicit = "implicit",
    Invoked = "invoked",
}
```
seems fine to me. Curious what you think the value of preemptively adding a `Speculative` option would be.

@mjbvz is there potential that VSCode might want to send info in addition to the `triggerReason`, such as a request for a specific "`refactorKind`"?

Reply:
My `Speculative` is your `Implicit`.